src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -622,80 +629,128 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
             _incomingBuffer.Discard(frameHeader.Length);
         }
 
-        private async ValueTask AcquireWriteLockAsync()
+        private async ValueTask StartWriteAsync(int credit, CancellationToken cancellationToken = default, bool acquireWriteLock = true)
         {
-            await _writerLock.WaitAsync().ConfigureAwait(false);
-
-            // If the connection has been aborted, then fail now instead of trying to send more data.
-            if (IsAborted())
+            if (acquireWriteLock)
             {
-                throw new IOException(SR.net_http_invalid_response);
+                await AcquireWriteLockAsync(cancellationToken).ConfigureAwait(false);
             }
-        }
 
-        private void ReleaseWriteLock()
-        {
-            // Currently, we always flush the write buffer before releasing the lock.
-            // If we change this in the future, we will need to revisit this assert.
-            Debug.Assert(_outgoingBuffer.ActiveMemory.IsEmpty);
+            try
+            {
+                // If there is a pending write that was canceled while in progress, wait for it to complete.
+                if (_inProgressWrite != null)
+                {
+                    await _inProgressWrite;
+                    _inProgressWrite = null;
+                }
+
+                int totalBufferLength = _outgoingBuffer.TotalLength;
+                int activeBufferLength = _outgoingBuffer.ActiveSpan.Length;
 
-            _writerLock.Release();
+                if (totalBufferLength >= 32768 &&
```

Comment:
We should never have "dead" space at the beginning of the buffer here. The pattern should always be: Fill the buffer until we are ready to flush it, then flush the entire buffer (up to the end of the active bytes), then discard all active bytes, meaning the buffer is now empty and we start writing at the beginning again.

This is somewhat obscured by the fact that we use a single structure (ArrayBuffer) for both read buffer and write buffer. In retrospect I'm not sure this is a great idea. 

I don't think this actually changes the code here, just FYI.

Reply:
I can add a comment explaining this in the code, if you think this needs some additional clarity. Otherwise I'll leave it as-is.

Reply:
It's fine. More just food for thought, in case we decide to simplify the buffer management here.